pkg/networkserver/grpc_deviceregistry.go

```diff
@@ -275,392 +269,1752 @@ func (ns *NetworkServer) Get(ctx context.Context, req *ttnpb.GetEndDeviceRequest
+			"mac_state.current_parameters.rejoin_count_periodicity",
+			"mac_state.current_parameters.rejoin_time_periodicity",
+			"mac_state.current_parameters.rx1_data_rate_offset",
+			"mac_state.current_parameters.rx1_delay",
+			"mac_state.current_parameters.rx2_data_rate_index",
+			"mac_state.current_parameters.rx2_frequency",
+			"mac_state.current_parameters.uplink_dwell_time",
+			"mac_state.desired_parameters.adr_ack_delay_exponent.value",
+			"mac_state.desired_parameters.adr_ack_limit_exponent.value",
+			"mac_state.desired_parameters.adr_data_rate_index",
+			"mac_state.desired_parameters.adr_nb_trans",
+			"mac_state.desired_parameters.adr_tx_power_index",
+			"mac_state.desired_parameters.beacon_frequency",
+			"mac_state.desired_parameters.channels",
+			"mac_state.desired_parameters.downlink_dwell_time",
+			"mac_state.desired_parameters.max_duty_cycle",
+			"mac_state.desired_parameters.max_eirp",
+			"mac_state.desired_parameters.ping_slot_data_rate_index_value.value",
+			"mac_state.desired_parameters.ping_slot_frequency",
+			"mac_state.desired_parameters.rejoin_count_periodicity",
+			"mac_state.desired_parameters.rejoin_time_periodicity",
+			"mac_state.desired_parameters.rx1_data_rate_offset",
+			"mac_state.desired_parameters.rx1_delay",
+			"mac_state.desired_parameters.rx2_data_rate_index",
+			"mac_state.desired_parameters.rx2_frequency",
+			"mac_state.desired_parameters.uplink_dwell_time",
+			"mac_state.device_class",
+			"mac_state.last_confirmed_downlink_at",
+			"mac_state.last_dev_status_f_cnt_up",
+			"mac_state.last_downlink_at",
+			"mac_state.last_network_initiated_downlink_at",
+			"mac_state.lorawan_version",
+			"mac_state.ping_slot_periodicity.value",
+			"mac_state.queued_responses",
+			"mac_state.recent_uplinks",
+			"mac_state.rejected_adr_data_rate_indexes",
+			"mac_state.rejected_adr_tx_power_indexes",
+			"mac_state.rejected_data_rate_ranges",
+			"mac_state.rejected_frequencies",
+			"mac_state.rx_windows_available",
+			"multicast",
 			"session.dev_addr",
-			"session.keys.f_nwk_s_int_key.key",
-		); err != nil {
-			return nil, nil, errInvalidFieldMask.WithCause(err)
-		}
-		req.EndDevice.DevAddr = &req.EndDevice.Session.DevAddr
-		sets = ttnpb.AddFields(sets,
-			"ids.dev_addr",
+			"session.last_conf_f_cnt_down",
+			"session.last_f_cnt_up",
+			"session.last_n_f_cnt_down",
+			"session.queued_application_downlinks",
+			"supports_join",
 		)
+	}
 
-		if req.EndDevice.LoRaWANVersion.Compare(ttnpb.MAC_V1_1) >= 0 {
-			if err := ttnpb.RequireFields(sets,
-				"session.keys.nwk_s_enc_key.key",
-				"session.keys.s_nwk_s_int_key.key",
-			); err != nil {
-				return nil, nil, errInvalidFieldMask.WithCause(err)
+	var evt events.Event
+	dev, ctx, err := ns.devices.SetByID(ctx, st.Device.EndDeviceIdentifiers.ApplicationIdentifiers, st.Device.EndDeviceIdentifiers.DeviceID, st.GetFields(), st.SetFunc(func(ctx context.Context, stored *ttnpb.EndDevice) error {
```

Comment:
In its current state, this fails for devices without a session with the following error:

```
 ERROR Failed to set device in registry         error=error:pkg/redis:store (store error) error_cause='ping_slot_frequency' has no subfields, but [value] were specified grpc.method=Set grpc.service=ttn.lorawan.v3.NsEndDeviceRegistry namespace=grpc peer.address=[::1]:55570 peer.real_ip=::1 request_id=01EXTCNTRKTPFPG16B8HPA28NR

```

After investigating a little, this is returned in https://github.com/rvolosatovs/lorawan-stack/blob/edc7d17780db04fc8fbfa252467cbabc8cf7bd85/pkg/networkserver/redis/registry.go#L726

because `gets` contains `mac_settings.ping_slot_frequency.value`, but `updated.MACSettings.PingSlotFrequency` is `nil`, and the pb setter does not like that. 

Any fix may also need to be applied in lines 541 and 563 as well.

Reply:
Can you provide a device model I can reproduce it with? I'm able to create devices with no session just fine locally.

Reply:
This is an example that fails:

```
{"ids":{"device_id":"uno","application_ids":{"application_id":"akolaitis-testing"},"dev_eui":"0101010101010101","join_eui":"0101010101010101"},"created_at":"0001-01-01T00:00:00Z","updated_at":"0001-01-01T00:00:00Z","name":"uno","lorawan_version":"1.0.2","lorawan_phy_version":"1.0.2-b","frequency_plan_id":"EU_863_870_TTN","supports_join":true,"root_keys":{"app_key":{"key":"01010101010101010101010101010101"}},"mac_settings":{"supports_32_bit_f_cnt": true}}
```

It works if no MAC settings are set:

```
{"ids":{"device_id":"uno","application_ids":{"application_id":"akolaitis-testing"},"dev_eui":"0101010101010101","join_eui":"0101010101010101"},"created_at":"0001-01-01T00:00:00Z","updated_at":"0001-01-01T00:00:00Z","name":"uno","lorawan_version":"1.0.2","lorawan_phy_version":"1.0.2-b","frequency_plan_id":"EU_863_870_TTN","supports_join":true,"root_keys":{"app_key":{"key":"01010101010101010101010101010101"}},"mac_settings":{"supports_32_bit_f_cnt": true}}
```